File: 24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #prev-->p-->q-->Next  to 
        #prev-->q-->p-->Next
        cur=head        
        prev=ListNode()
        head1=prev
        while cur:
            if cur.next:
                p=cur
                q=cur.next
                Next=q.next
                prev.next=q                
                q.next=p
                p.next=Next
                cur=Next
                prev=p 
            else:
                prev.next=cur
                cur=cur.next        
        return head1.next

File: 24. Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {

    ListNode* cur=head;   
    ListNode *prev=new ListNode();    
    prev->next=head;    
    ListNode *dummy=new ListNode();
    dummy=prev;     
    while (cur!=NULL and cur->next!=NULL ){
        ListNode* temp=cur;
        ListNode* nextPair=cur->next->next;       
        prev->next=cur->next;
        cur->next->next=temp;
        temp->next=nextPair;
        prev=temp;
        cur=nextPair;      
    }        
     return (dummy->next);   
    }
};
